mne_bids/read.py

```diff
@@ -369,18 +382,58 @@ def read_raw_bids(bids_fname, bids_root, extra_params=None,
     # Session is optional ...
     if params['ses'] is not None:
         kind_dir = op.join(kind_dir, 'ses-{}'.format(params['ses']))
-    # Kind is mandatory
-    kind_dir = op.join(kind_dir, kind)
+    # Kind is mandatory. Check which kinds are available for this particular
+    # subject & session. If we get no or multiple hits, throw an error.
+    kinds = _get_kinds_for_sub(bids_basename=bids_basename,
+                               bids_root=bids_root,
+                               sub=params['sub'],
+                               ses=params['ses'])
+    if not kinds:
+        raise ValueError('No data found.')
+    elif kind is None and len(kinds) == 1:
+        kind = kinds[0]
+    elif kind is None and len(kinds) >= 2:
+        msg = (f'Found data of more than one recording modality. Please pass '
+               f'the `kind` parameter to specify which data to load. Found '
+               f'the following kinds: {kinds}')
+        raise RuntimeError(msg)
 
-    config = None
-    if ext in ('.fif', '.ds', '.vhdr', '.edf', '.bdf', '.set', '.sqd', '.con'):
-        bids_fpath = op.join(kind_dir,
-                             bids_basename + '_{}{}'.format(kind, ext))
+    kind_dir = op.join(kind_dir, kind)
 
-    elif ext == '.pdf':
-        bids_raw_folder = op.join(kind_dir, bids_basename + '_{}'.format(kind))
+    # Find all matching files in all supported formats.
+    valid_exts = ('.fif', '.ds', '.vhdr', '.edf', '.bdf', '.set', '.sqd',
+                  '.con', '.pdf')
+
+    matching_paths = glob.glob(op.join(kind_dir,
+                                       f'{bids_basename}_{kind}' + '.*'))
+    matching_paths = [p for p in matching_paths
+                      if op.splitext(p)[1] in valid_exts]
+
+    if not matching_paths:
+        msg = ('Could not locate a data file of a supported format. This is '
+               'likely a problem with your BIDS dataset. Please run the BIDS '
+               'validator on your data.')
+        raise RuntimeError(msg)
+    # FIXME This will break e.g. with FIFF data split across multiple files
+    if len(matching_paths) > 1:
+        msg = ('Found more than one matching data file for the requested '
+               'recording. Cannot proceed due to the ambiguity. This is '
+               'likely a problem with your BIDS dataset. Please run the BIDS '
+               'validator on your data.')
+        raise RuntimeError(msg)
+
+    matching_path = matching_paths[0]
```

Comment:
I think this should go in a separate function once you've figured out what makes it work. There are also [some helper functions already in there](https://github.com/mne-tools/mne-bids/blob/25f5b433fdf070ee0942276e8b271523e29bf819/mne_bids/utils.py#L575), so you might want to check if you can reduce duplication.

Reply:
I looked and I think for this specific case, the existing function(s) unfortunately don't provide much help